diff -urN package/acl.ORIG/acl.mk package/acl/acl.mk
--- package/acl.ORIG/acl.mk	2013-01-29 19:11:34.365379429 -0800
+++ package/acl/acl.mk	2013-01-29 20:25:02.369412996 -0800
@@ -45,7 +45,7 @@
 	touch $@
 
 $(TARGET_DIR)/$(ACL_TARGET_BINARY): $(ACL_DIR)/.built
-	$(MAKE) CC=$(TARGET_CC) prefix=$(STAGING_DIR)/usr -C $(ATTR_DIR) install-lib install-dev
+	$(MAKE) CC=$(TARGET_CC) prefix=$(STAGING_DIR)/usr -C $(ATTR_DIR) install install-lib install-dev
 	$(INSTALL) -m 0755 $(STAGING_DIR)/usr/bin/chacl $(TARGET_DIR)/usr/bin/
 	$(INSTALL) -m 0755 $(STAGING_DIR)/usr/bin/setfacl $(TARGET_DIR)/usr/bin/
 	$(INSTALL) -m 0755 $(STAGING_DIR)/usr/bin/getfacl $(TARGET_DIR)/usr/bin/
