--- package/file/file.mk.ORIG	2011-07-31 20:34:04.000000000 -0600
+++ package/file/file.mk	2011-07-31 20:34:38.000000000 -0600
@@ -3,7 +3,7 @@
 # file
 #
 #############################################################
-FILE_VERSION:=4.24
+FILE_VERSION:=4.26
 FILE_SOURCE:=file_$(FILE_VERSION).orig.tar.gz
 FILE_SITE:=$(BR2_DEBIAN_MIRROR)/debian/pool/main/f/file
 FILE_SOURCE_DIR:=$(BUILD_DIR)/file-$(FILE_VERSION)
