diff -urN package/proftpd.ORIG/proftpd.mk package/proftpd/proftpd.mk
--- package/proftpd.ORIG/proftpd.mk	2013-02-12 15:40:13.461632529 -0800
+++ package/proftpd/proftpd.mk	2013-02-19 23:00:58.969514367 -0800
@@ -59,7 +59,7 @@
 		$(DISABLE_LARGEFILE) \
 		$(ENABLE_IPV6) \
 		--with-gnu-ld \
-		--with-modules=mod_sql:mod_sql_sqlite:mod_vroot \
+		--with-modules=mod_sql:mod_sql_sqlite:mod_dbacl:mod_vroot \
 	)
 	touch $@
 
