diff -urN package/proftpd.ORIG/proftpd.mk package/proftpd/proftpd.mk
--- package/proftpd.ORIG/proftpd.mk	2013-02-12 00:28:52.652377630 -0800
+++ package/proftpd/proftpd.mk	2013-02-12 00:31:26.670416286 -0800
@@ -77,7 +77,7 @@
 	$(INSTALL) -m 0755 package/proftpd/S50proftpd $(TARGET_DIR)/etc/init.d
 	$(INSTALL) -m 0755 $(PROFTPD_DIR)/ftpwho $(TARGET_DIR)/usr/sbin  ###WAGLE###
 
-proftpd: uclibc $(TARGET_DIR)/$(PROFTPD_TARGET_BINARY)
+proftpd: uclibc sqlite $(TARGET_DIR)/$(PROFTPD_TARGET_BINARY)
 
 proftpd-clean:
 	rm -f $(TARGET_DIR)/$(PROFTPD_TARGET_BINARY)
