diff -urN package/proftpd.ORIG/proftpd.mk package/proftpd/proftpd.mk
--- package/proftpd.ORIG/proftpd.mk	2013-02-19 23:13:17.762487660 -0800
+++ package/proftpd/proftpd.mk	2013-02-20 10:49:17.723119645 -0800
@@ -16,6 +16,11 @@
 PROFTPD_MOD_VROOT_SITE:=http://www.castaglia.org/proftpd/modules/
 PROFTPD_MOD_VROOT_CAT:=$(ZCAT)
 
+PROFTPD_MOD_DBACL_VERSION:=0.0
+PROFTPD_MOD_DBACL_SOURCE:=proftpd-mod-dbacl-$(PROFTPD_MOD_DBACL_VERSION).tar.gz
+PROFTPD_MOD_DBACL_SITE:=http://www.castaglia.org/proftpd/modules/
+PROFTPD_MOD_DBACL_CAT:=$(ZCAT)
+
 ifeq ($(BR2_INET_IPV6),y)
 ENABLE_IPV6:=--enable-ipv6
 else
@@ -23,16 +28,21 @@
 endif
 
 $(DL_DIR)/$(PROFTPD_SOURCE):
-	 $(WGET) -P $(DL_DIR) $(PROFTPD_SITE)/$(PROFTPD_SOURCE)
+	$(WGET) -P $(DL_DIR) $(PROFTPD_SITE)/$(PROFTPD_SOURCE)
 
 $(DL_DIR)/$(PROFTPD_MOD_VROOT_SOURCE):
-	 $(WGET) -P $(DL_DIR) $(PROFTPD_MOD_VROOT_SITE)/$(PROFTPD_MOD_VROOT_SOURCE)
+	$(WGET) -P $(DL_DIR) $(PROFTPD_MOD_VROOT_SITE)/$(PROFTPD_MOD_VROOT_SOURCE)
+
+$(DL_DIR)/$(PROFTPD_MOD_DBACL_SOURCE):
+	$(WGET) -P $(DL_DIR) $(PROFTPD_MOD_DBACL_SITE)/$(PROFTPD_MOD_DBACL_SOURCE)
+
 
-proftpd-source: $(DL_DIR)/$(PROFTPD_SOURCE) $(DL_DIR)/$(PROFTPD_MOD_VROOT_SOURCE)
+proftpd-source: $(DL_DIR)/$(PROFTPD_SOURCE) $(DL_DIR)/$(PROFTPD_MOD_VROOT_SOURCE)  $(DL_DIR)/$(PROFTPD_MOD_DBACL_SOURCE)
 
 $(PROFTPD_DIR)/.unpacked: $(DL_DIR)/$(PROFTPD_SOURCE)
 	$(PROFTPD_CAT) $(DL_DIR)/$(PROFTPD_SOURCE) | tar -C $(BUILD_DIR) $(TAR_OPTIONS) -
 	$(PROFTPD_MOD_VROOT_CAT) $(DL_DIR)/$(PROFTPD_MOD_VROOT_SOURCE) | tar -C $(PROFTPD_DIR)/contrib --strip-components 1 $(TAR_OPTIONS) - mod_vroot/mod_vroot.c
+	$(PROFTPD_MOD_DBACL_CAT) $(DL_DIR)/$(PROFTPD_MOD_DBACL_SOURCE) | tar -C $(PROFTPD_DIR)/contrib --strip-components 1 $(TAR_OPTIONS) - mod_dbacl/mod_dbacl.c
 	$(CONFIG_UPDATE) $(PROFTPD_DIR)
 	toolchain/patch-kernel.sh $(PROFTPD_DIR) package/proftpd/ proftpd-$(PROFTPD_VERSION)\*.patch;
 	touch $@
