--- proftpd-1.3.4b/contrib/mod_vroot.c.ORIG	2011-01-10 18:41:44.000000000 -0800
+++ proftpd-1.3.4b/contrib/mod_vroot.c	2013-02-20 22:22:11.096733678 -0800
@@ -1279,7 +1279,11 @@
       "' is not an absolute path", NULL));
   }
 
-  c = add_config_param_str(cmd->argv[0], 2, cmd->argv[1], cmd->argv[2]);
+  // ###WAGLE###
+  c = add_config_param_str(cmd->argv[0], 2
+    , path_subst_uservar(cmd->tmp_pool,&cmd->argv[1])
+    , path_subst_uservar(cmd->tmp_pool,&cmd->argv[2])
+    );
 
   /* Set this flag in order to allow mod_ifsession to work properly with
    * multiple VRootAlias directives.
