--- linux-kernel/drivers/usb/storage/scsiglue.c.ORIG	2012-01-09 21:26:03.000000000 -0700
+++ linux-kernel/drivers/usb/storage/scsiglue.c	2012-01-09 21:27:04.000000000 -0700
@@ -275,7 +275,7 @@
 	 * Let the scanning code know if this target merely sets
 	 * Peripheral Device Type to 0x1f to indicate no LUN.
 	 */
-	if (us->subclass == USB_SC_UFI)
+	if (us->subclass == US_SC_UFI)
 		starget->pdt_1f_for_no_lun = 1;
 
 	return 0;
